geo/src/algorithm/stitch.rs

```diff
@@ -0,0 +1,582 @@
+    lines: Vec<Line<F>>,
+) -> PolygonStitchingResult<MultiPolygon<F>> {
+    let rings = stitch_rings_from_lines(lines)?;
+
+    fn find_parent_idxs<F: GeoFloat>(
+        ring_idx: usize,
+        ring: &LineString<F>,
+        all_rings: &[LineString<F>],
+    ) -> Vec<usize> {
+        all_rings
+            .iter()
+            .enumerate()
+            .filter(|(other_idx, _)| ring_idx != *other_idx)
+            .filter_map(|(idx, maybe_parent)| {
+                Polygon::new(maybe_parent.clone(), vec![])
+                    .contains(ring)
+                    .then_some(idx)
+            })
+            .collect()
+    }
+
+    // Associates every ring with its parents (the rings that contain it)
+    let parents: HashMap<usize, Vec<usize>> = rings
+        .iter()
+        .enumerate()
+        .map(|(ring_idx, ring)| {
+            let parent_idxs = find_parent_idxs(ring_idx, ring, &rings);
+            (ring_idx, parent_idxs)
+        })
+        .collect();
+
+    // Associates outer rings with their inner rings
+    let mut polygons_idxs: HashMap<usize, Vec<usize>> = HashMap::default();
+
+    // For each ring, we check how many parents it has  otherwise it's an outer ring
+    //
+    // This is important in the scenarios of "donuts" where you have an outer donut shaped
+    // polygon which completely contains a smaller polygon inside its hole
+    for (ring_index, parent_idxs) in parents.iter() {
+        let parent_count = parent_idxs.len();
+
+        // if it has an even number of parents, it's an outer ring so we can just add it if it's
+        // missing
+        if parent_count % 2 == 0 {
+            polygons_idxs.entry(*ring_index).or_default();
+            continue;
+        }
+
+        // the direct parent is the parent which has itself the most parents
+        fn find_direct_parents(
+            parents_indexs: &[usize],
+            parents: &HashMap<usize, Vec<usize>>,
+        ) -> Option<usize> {
+            parents_indexs
+                .iter()
+                .filter_map(|parent_idx| {
+                    parents
+                        .get(parent_idx)
+                        .map(|grandparents| (parent_idx, grandparents))
+                })
+                .max_by_key(|(_, grandparents)| grandparents.len())
+                .map(|(idx, _)| idx)
+                .copied()
+        }
+
+        // if it has an odd number of parents, it's an inner ring
+        //
+        // to find the specific outer ring it is related to, we search for the direct parent.
+        if let Some(direct_parent) = find_direct_parents(parent_idxs, &parents) {
+            polygons_idxs
+                .entry(direct_parent)
+                .or_default()
+                .push(*ring_index);
+        }
+    }
+
+    // lookup rings by index and create polygons
+    let polygons = polygons_idxs
+        .into_iter()
+        .map(|(parent_idx, children_idxs)| {
+            let exterior = rings[parent_idx].clone();
```

Comment:
perf: These clones are unfortunate. 

Do we expect `rings[i]` for a given `i` to be used in more than one place? It seems like we should be able to `take` or something to avoid the clones.

Reply:
nothing that immediately jumps to mind. The issue with this again is that:

- a) the compiler is smarter than both of us together and already does a good job of optimizing this
- b) any preparations that are needed to make a `take` possible are costing us perf gains (that is, if any exist)

I'd leave it at this for now